src/Orleans/Core/Grain.cs

```diff
@@ -39,7 +36,7 @@ protected IGrainFactory GrainFactory
         /// </summary>
         protected IServiceProvider ServiceProvider 
         {
-            get { return Runtime?.ServiceProvider; }
+            get { return Data.ServiceProvider ?? Runtime?.ServiceProvider; }
```

Comment:
> Still, it doesn't seem unsafe, does it?

I think it's safe, but it's difficult to verify that this will always be safe, which is why I suggest the failure option. Additionally, we should probably remove `ServiceProvider` from `IGrainRuntime` since it's no longer valid: IGrainRuntime is designed for grains, but its `ServiceProvider` has a different scope.